Add controllers

egornowik21 opened the pull request.

egornowik21 committed:
Добавление слоя сервиса и контроллеров для юзера

egornowik21 committed:
добавил альтернативную версию для метода patch, но все равно падает с ошибкой

egornowik21 committed:
User создается и обновляется без ошибок

egornowik21 committed:
Добавлено разделение хранилища и сервиса и соблюден чекстайл

egornowik21 committed:
Версия с пройденными тестами Postman v1

egornowik21 committed:
Версия с пройденными тестами Postman v2

pattyku commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,34 @@
+    String description;
+    @NotNull(message = "Стаус не может быть пустым")
+    Boolean available;
+    User owner;
```
Тут отлично, что используешь объекты User, Request ..., а не их айди. Многие сделали именно так

egornowik21 replied:
Да, спасибо) так сразу к объекту можно обратиться

pattyku commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,34 @@
+    @NotBlank(message = "Имя не может быть пустым")
+    @NotNull(message = "Имя не может быть пустым")
+    @NotEmpty(message = "Описание не может быть пустым")
+    String name;
```
Для строк используем аннотацию `@NotBlank`, аннотация включает проверку на null
Хорошая статься про различия данных аннотаций https://www.baeldung.com/java-bean-validation-not-null-empty-blank

egornowik21 replied:
а остальные? я их для профилактики добавил
статью прочитаю

egornowik21 replied:
в методы с patch не стал добавлять валидацию через аннотации, так как там же может прийти запрос с почтой =null или именем = null
это же частичное обновление сущности
оставил свои старые методы для этих эндпоинтов

pattyku replied:
> а остальные? я их для профилактики добавил статью прочитаю

Для строк мы используем аннотацию `@NotBlank` в основном. Исключение если строка у нас может быть пустой, но не может быть null  `@NotNull`
Если нам нужно чтоб 
```
@NotBlank(message = "Имя не может быть пустым")
String name;
```

 @NotEmpty используем для коллекций. Мапы, листы и т.д.

pattyku replied:
> в методы с patch не стал добавлять валидацию через аннотации, так как там же может прийти запрос с почтой =null или именем = null это же частичное обновление сущности оставил свои старые методы для этих эндпоинтов

тут все верно

pattyku commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,49 @@
+@RequiredArgsConstructor
 public class ItemController {
+
+    private final ItemService itemService;
```
хорошо бы добавить логи в каждый метод
`log.info("POST /items - создание новой вещи.");`

egornowik21 replied:
добавил, но раньше я это делал на уровне сервиса. теперь сделал на уровне контроллера
правильно я понимаю, что логи с info идут в контроллере
а логи в ошибками (error) в сервисе?

pattyku replied:
Входные и выходные данные лучше логгировать в контроллере. 
В сервисе можно логгировать не только ошибки, но и сложную логику. 
Чаще всего в info выводят критичную информацию, не всегда на проекте выделяется много ресурсов для хранения логов. Поэтому в контроллерах иногда принимают решение логгировать информацию  на уровне debug и включать по необходимости. 
Четких и жестких правил нет. Многое зависит от самого проекта, ресурсов и принятых практик в команде

pattyku commented on src/main/java/ru/practicum/shareit/item/dao/ItemDaoImpl.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    public List<ItemDto> getAllItems() {
+        ArrayList<ItemDto> itemDtoList = new ArrayList<>();
+        for (Long itemId : items.keySet()) {
```
может попробуем цикл заменить на стримы
Напимер здесь можно сделать так
Пример (не подходит копипаст)

```
    public List<ItemDto> getAllItems(Integer owner) {
        validateUser(owner);
        List<Item> itemList = itemRepository.getAllItems(owner);
        log.info("Получен список всех вещей.");

        return itemList
                .stream()
                .map(ItemMapper::toItemDto)
                .collect(Collectors.toList());
    }
```

Старайся почаще применять стримы. На работе будешь часто сталкиваться

egornowik21 replied:
заменил все циклы на стримы, если есть чем их дополнить - буду рад любым советам)

pattyku commented on src/main/java/ru/practicum/shareit/item/dao/ItemDaoImpl.java:
```diff
@@ -0,0 +1,138 @@
+    public List<ItemDto> searchItem(String text) {
+        ArrayList<ItemDto> itemDtoList = new ArrayList<>();
+        String textInput = text.toLowerCase();
+        if (text != null && !(text.isEmpty()) && !(text.isBlank())) {
```
Здесь лучше сделать обратную проверку (if (text == null и т.д.) и лучше сделать это сразу же. Сейчас вначале мы создаем пустую коллекцию. и переменную которые нам не понадобятся если проверка ниже не пройдет. 
А еще лучше перенести проверку на уровень котроллера. Зачем вызывать сервис, а потом еще слой дао если текст пустой или равен null. 
Пример 
```
	@GetMapping("/search")
	public List<ItemDto> search(@RequestParam(name = "text") String text) {
		log.info("Search item by text={}", text);
		if (text == null || text.isBlank()) {
			return Collections.emptyList();
		}

		return itemService.searchItems(text);
	}
```

Давай вынесем проверку на уровень контроллера

egornowik21 replied:
согласен, сделал
а отлавливание исключений, где пользователь не найден лучше в слое контроллера сделать или же на уровне хранилища оставить?

pattyku replied:
на уровне сервиса
```
		final User user = userDao.findById(userId);
		if (user == null) {
			throw new NotFoundException("User with id=" + newUser.getId() + " not found");
		}
```

pattyku commented on src/main/java/ru/practicum/shareit/user/dao/UserDaoImpl.java:
```diff
@@ -0,0 +1,95 @@
+        UserDto usetToUpdate = getUserbyId(id);
+        String name = userDto.getName();
+        String mail = userDto.getEmail();
+        if (mail != null) {
```
Для проверки уникальности можно использовать сет
private final Set<String> emailUniqSet = new HashSet<>();

egornowik21 replied:
и здесь тоже добавил хэшсет

egornowik21 committed:
Первые исправления после первого ревью

egornowik21 committed:
Вторые исправления после первого ревью

egornowik21 committed:
Финальные исправления после первого ревью
- добавлены стримы
- добавлены аннотации
- добавлен сет для меэйлов пользователей
- добавлены логи в контроллере
- измнена проверка текста для эндпоинта поиска

egornowik21 reviewed (commented):
новые изменения

egornowik21 committed:
Финальные исправления после первого ревью
- добавлены стримы
- добавлены аннотации
- добавлен сет для меэйлов пользователей
- добавлены логи в контроллере
- измнена проверка текста для эндпоинта поиска

egornowik21 committed:
Финальные исправления после первого ревью v3 - checkstyle
- добавлены стримы
- добавлены аннотации
- добавлен сет для меэйлов пользователей
- добавлены логи в контроллере
- измнена проверка текста для эндпоинта поиска

egornowik21 committed:
Финальные исправления после первого ревью v4 - checkstyle
- добавлены стримы
- добавлены аннотации
- добавлен сет для меэйлов пользователей
- добавлены логи в контроллере
- измнена проверка текста для эндпоинта поиска

egornowik21 committed:
Финальные исправления после первого ревью v5 - checkstyle
- добавлены стримы
- добавлены аннотации
- добавлен сет для меэйлов пользователей
- добавлены логи в контроллере
- измнена проверка текста для эндпоинта поиска
